In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
pd.set_option('display.max_columns', None)
sns.set_style("whitegrid")
import warnings
warnings.filterwarnings("ignore")
from sklearn.model_selection import train_test_split

In [30]:
df = pd.read_csv('Airline_Final_Classification.csv')
df.head()

DEP_DELAY  tavg  prcp  wspd  OP_CARRIER_AS  OP_CARRIER_B6  OP_CARRIER_DL  \
0          0  18.3   0.0  12.6          False          False          False   
1          0  18.3   0.0  12.6          False          False          False   
2          1  18.3   0.0  12.6          False          False          False   
3          0  18.3   0.0  12.6          False          False          False   
4          0  18.3   0.0  12.6          False          False          False   

   OP_CARRIER_EV  OP_CARRIER_F9  OP_CARRIER_HA  OP_CARRIER_MQ  OP_CARRIER_NK  \
0          False          False          False          False           True   
1          False          False          False          False           True   
2          False          False          False          False           True   
3          False          False          False          False           True   
4          False          False          False          False           True   

   OP_CARRIER_OO  OP_CARRIER_UA  OP_CARRIER_US  OP_CARRIER_VX  OP_CARRIER_WN  \
0          False          False          False          False          False   
1          False          False          False          False          False   
2          False          False          False          False          False   
3          False          False          False          False          False   
4          False          False          False          False          False   

   HOUR_1  HOUR_2  HOUR_3  HOUR_4  HOUR_5  HOUR_6  HOUR_7  HOUR_8  HOUR_9  \
0   False   False   False   False   False   False   False   False   False   
1   False   False   False   False   False   False   False   False   False   
2   False   False   False   False   False   False   False    True   False   
3   False   False   False   False   False   False   False   False   False   
4   False   False   False   False   False   False   False   False   False   

   HOUR_10  HOUR_11  HOUR_12  HOUR_13  HOUR_14  HOUR_15  HOUR_16  HOUR_17  \
0    False    False    False    False    False    False    False    False   
1     True    False    False    False    False    False    False    False   
2    False    False    False    False    False    False    False    False   
3    False    False    False    False    False    False    False    False   
4    False    False    False    False    False    False    False    False   

   HOUR_18  HOUR_19  HOUR_20  HOUR_21  HOUR_22  HOUR_23  
0    False    False    False     True    False    False  
1    False    False    False    False    False    False  
2    False    False    False    False    False    False  
3     True    False    False    False    False    False  
4    False    False     True    False    False    False

In [31]:
from sklearn.model_selection import train_test_split
y = df['DEP_DELAY']
X = df.drop('DEP_DELAY', axis=1)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [6]:
# from sklearn.preprocessing import StandardScaler

# sc = StandardScaler()
# X_train_std = sc.fit_transform(X_train)
# X_test_std = sc.transform(X_test)

In [32]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report,confusion_matrix,f1_score,roc_auc_score,roc_curve,accuracy_score
from sklearn.model_selection import cross_val_score
LR = LogisticRegression(random_state=0,solver="liblinear")
LR.fit(X_train, y_train)
y_pred = LR.predict(X_test)
print('Accuracy:', accuracy_score(y_pred,y_test))
print('confusion matrix-test\n',confusion_matrix(y_pred,y_test))
cv_scores = cross_val_score(LR, X, y, cv=5)
print()
print(classification_report(y_test, y_pred))
print()
print("Average 5-Fold CV Score: {}".format(round(np.mean(cv_scores),4)),
      ", Standard deviation: {}".format(round(np.std(cv_scores),4)))

Accuracy: 0.6598804297002508
confusion matrix-test
 [[933957 442971]
 [135319 188008]]

              precision    recall  f1-score   support

           0       0.68      0.87      0.76   1069276
           1       0.58      0.30      0.39    630979

    accuracy                           0.66   1700255
   macro avg       0.63      0.59      0.58   1700255
weighted avg       0.64      0.66      0.63   1700255


Average 5-Fold CV Score: 0.6557 , Standard deviation: 0.0102


1. Accuracy
Accuracy (0.6599): This is the proportion of total predictions (for delayed and not delayed flights) that were correct. Your model correctly predicts whether a flight will be delayed or not about 66% of the time. While this is better than random guessing, considering the complexity of the factors influencing flight delays, there's room for improvement.
2. Confusion Matrix
The confusion matrix is a table used to describe the performance of a classification model on a set of test data for which the true values are known. It breaks down the predictions into four categories:

True Negatives (TN): 933,957 - The model correctly predicted the flights that were not delayed.
<br>
False Positives (FP): 442,971 - The model incorrectly predicted these flights would be delayed, but they were not.
<br>
False Negatives (FN): 135,319 - The model incorrectly predicted these flights would not be delayed, but they were.
<br>
True Positives (TP): 188,008 - The model correctly predicted the flights that were delayed.
<br>
<br>
3. Precision, Recall, F1-Score for Each Class
<br><br>
Precision (0 for non-delayed flights: 0.68, 1 for delayed flights: 0.58): For non-delayed flights, the model has a higher precision, indicating it's more reliable in predicting flights that are not delayed compared to predicting delays.
<br><br>
Recall (0 for non-delayed flights: 0.87, 1 for delayed flights: 0.30): The model has high recall for non-delayed flights, meaning it's good at catching non-delay cases but not as effective at identifying actual delays.
<br><br>
F1-Score (0 for non-delayed flights: 0.76, 1 for delayed flights: 0.39): The higher F1 score for non-delayed flights indicates better performance in predicting non-delayed flights over delayed ones.
<br><br>
4. Macro Average and Weighted Average
Macro Average Precision, Recall, F1-Score (0.63, 0.59, 0.58 respectively): This averages the precision, recall, and F1-score for each class without taking their occurrence into account. It treats all classes equally, emphasizing the performance on the minority class.
Weighted Average Precision, Recall, F1-Score (0.64, 0.66, 0.63 respectively): This accounts for class imbalance by weighting the average based on the number of instances in each class. It provides a measure of quality that reflects the true distribution of the classes in the data.
<br><br>
5. Average 5-Fold CV Score
Average 5-Fold CV Score (0.6557) with Standard Deviation (0.0102): This shows the average accuracy obtained across five different splits of the dataset into training and testing sets, along with the standard deviation of these accuracies. A lower standard deviation indicates that the model's performance is consistent across different subsets of the data.
<br><br>

##### Conclusion
The logistic regression model shows moderate performance in predicting flight delays with an accuracy of around 66%. It performs better at predicting non-delayed flights than delayed ones, as indicated by higher precision, recall, and F1-scores for the non-delayed class. The model's consistency is confirmed by a relatively stable average 5-Fold CV score. However, the lower recall and F1-score for the delayed class suggest there's significant room for improvement, particularly in correctly identifying delayed flights. Having access to some features like different sets of delays, size of operations, or trying more complex models might improve performance.